### PR TITLE
[BugFix] Fix mishandled type null

### LINKED ISSUE #34991: [Bug] array_concat lead to be crash
### Steps to reproduce the behavior (Required)

```
with t0 as (
    select c1 from (values([])) as t(c1)
)
select 
array_concat(c1, [1])
from t0;

with t0 as (
    select c1 from (values([])) as t(c1)
)
select 
array_concat([1], c1)
from t0;
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by liuyehcf**
What I'm doing:
1. Fix mishanded type null, which may lead to be crash or report `Array<NULL_TYPE> is illegal in thrift stage`
2. Fix the erroneous null handling logic, which misses the cast operator, that lead to mismatch of column types.

Fixes #34991,https://github.com/StarRocks/starrocks/issues/34982

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java [hunk 1]
```diff
@@ -91,10 +91,10 @@ public ScalarOperator visitCall(CallOperator call, ScalarOperatorRewriteContext
...
                 ScalarOperator child = call.getChild(i);
 
-                //Cast from array(null), direct assignment type to avoid passing null_literal into be
+                // Cast from array(null), direct assignment type to avoid passing null_literal into be
                 if (type.isArrayType() && child.getType().isArrayType()
                         && ((ArrayType) child.getType()).getItemType().isNull()) {
-                    child.setType(type);
+                    addCastChild(type, call, i);
                     continue;
                 }
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java:99]: what about other nested types, map, struct,json?

**Review by fzhedu — COMMENTED** [review]
I wonder that why this bug happens with CTE, no CTE works OK

**Comment by imay** [comment]
You'd better add the reproduce case to the sql-test case suits.

**Comment by liuyehcf** [comment]
> I wonder that why this bug happens with CTE, no CTE works OK

I think it's not the CTE that cause this problem, this query can also lead to be crash

```sql
select 
array_concat(c1, [1])
from (select c1 from (values([])) as t(c1))t;
```

The problem comes from the empty array, we cannot determine its type at the first place, so we can only assign it to `Array<Null>`, and in some situations, this intermediate type has not been properly handled(replaced by `Array<Bool>`), thereby cause be to crash.

**Comment by liuyehcf** [comment]
> You'd better add the reproduce case to the sql-test case suits.

Added.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java (hunk 1 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/scalar/ImplicitCastRule.java]: updated

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/MapTypeTest.java [hunk 2]
```diff
@@ -41,6 +41,22 @@ public void testMapFunc() throws Exception { // get super common return type
...
+                "from t0;";
+        plan = getFragmentPlan(sql);
+        assertContains(plan, "map_concat(map{'a':1,'b':2}, CAST(2: c1 AS MAP<VARCHAR,TINYINT>))");
+
+        sql = "with t0 as (\n" +
+                "    select c1 from (values(map())) as t(c1)\n" +
+                ")\n" +
+                "select map_concat(c1, map('a',1, 'b',2))\n" +
+                "from t0;";
+        plan = getFragmentPlan(sql);
```
Comment (fzhedu) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/MapTypeTest.java:58]: add some cases for nest map/array, like [map()],map([],[]),[[[]]]

Comment (fzhedu) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/MapTypeTest.java:58]: map([],map())

**Comment by fzhedu** [comment]
> > I wonder that why this bug happens with CTE, no CTE works OK
> 
> I think it's not the CTE that cause this problem, the following query can also lead to be crash
> 
> ```sql
> select 
> array_concat(c1, [1])
> from (select c1 from (values([])) as t(c1))t;
> ```
> 
> The problem comes from the empty array, we cannot determine its type at the first place, so we can only assign it to `Array<Null>`, and in some situations, this intermediate type has not been properly handled(replaced by `Array<Bool>`), thereby cause be to crash.

the normal select may be processed by different code path?

**Comment by fzhedu** [comment]
can we add more checks to avoid such problems before pushing down to BE?

**Review by fzhedu — APPROVED** [review]

**Review by packy92 — APPROVED** [review]

**Comment by luohaha** [comment]
https://github.com/Mergifyio backport branch-3.1-cs

**Pull request merged**